<div style="display: flex; align-items: center; gap: 20px; font-family: Arial, sans-serif;">
    <div style="margin-right: 30px;">
        <img src="../Sources/Logo.png" width="300" style="border-radius: 10px;"/>
    </div>
    <div style="line-height: 1.5;">
        <strong style="font-size: 18px; display: block; margin-bottom: 5px;">
            <a href='https://t.me/M4m4ev_Alex' style="text-decoration: none; color: #1a73e8;">
                Репетитор по информатике | Мамаев Алексей
            </a>
        </strong>
        <span style="font-size: 16px; color: #333;">
            Разбор <a href='https://euler.jakumo.org/problem/10-slozenie-prostyx-cisel' style="text-decoration: none; color: #1a73e8;">
            10-й задачи</a> проекта Эйлера
        </span>
        <span style="font-size: 16px; color: #333;">
            <br><br><br><br><br><br>ТГК: <a href='https://t.me/M4m4ev_Alex' style="text-decoration: none; color: #1a73e8;">@M4m4ev_Alex</a><br>
        </span>
    </div>
</div>

## Задача 10

Сумма простых чисел меньше 10 равна 2 + 3 + 5 + 7 = 17

Необходимо найти сумму всех простых чисел, меньших 2 000 000.

Ответ: 142913828922

## Используемые библиотеки

В данном разборе используются стандартные и сторонние библиотеки Python.

- `perf_counter` — измерение времени выполнения алгоритмов.
- `isprime` — функция библиотеки `sympy` для проверки числа на простоту.

In [1]:
from time import perf_counter
from sympy import isprime

## 1) Последовательная проверка чисел на простоту

В данном решении последовательно перебираются все нечётные числа, меньшие двух миллионов.

Для каждого числа с помощью функции `isprime()` выполняется проверка на простоту. Если число является простым, оно прибавляется к общей сумме.

Число `2` обрабатывается отдельно, так как является единственным чётным простым числом.

### Алгоритм

1. Начальная сумма инициализируется числом `2`.
2. Последовательно перебираются все нечётные числа от `3` до `2 000 000`.
3. Для каждого числа вызывается функция `isprime()`.
4. Если число является простым, оно добавляется к общей сумме.
5. После завершения перебора возвращается найденная сумма.

### Особенности

- Используется встроенная функция `isprime()` из библиотеки `sympy`.
- Проверяются только нечётные числа.
- Код отличается простотой и хорошо читается.
- Каждое число проверяется независимо от остальных.
- При больших диапазонах такой подход становится достаточно медленным.

### Вывод

Последовательная проверка чисел на простоту является простым и понятным способом решения задачи. Однако необходимость отдельно проверять каждое число приводит к большому количеству вычислений при увеличении диапазона поиска.

In [2]:
def get_result_1(n: int) -> int:
    sm = 2
    for i in range(3, n, 2):
        if isprime(i):
            sm += i
    return sm

start = perf_counter()
print(get_result_1(2_000_000))
end = perf_counter()

print(f'{end - start:.5f}')

142913828922
0.99774


## 2) Решето Эратосфена

В данном решении используется **решето Эратосфена** — один из наиболее эффективных алгоритмов поиска всех простых чисел в заданном диапазоне.

Вместо последовательной проверки каждого числа на простоту алгоритм сразу исключает все составные числа, оставляя только простые. После завершения работы решета остаётся лишь вычислить сумму найденных простых чисел.

### Алгоритм

1. Создаётся массив, в котором первоначально все числа считаются простыми.
2. Числа `0` и `1` помечаются как составные.
3. Последовательно перебираются числа, начиная с `2`.
4. Для каждого найденного простого числа все его кратные помечаются как составные.
5. После завершения работы алгоритма выполняется суммирование всех чисел, оставшихся простыми.
6. Полученная сумма возвращается в качестве ответа.

### Особенности

- Используется решето Эратосфена.
- Каждое составное число исключается только один раз.
- Отдельная проверка каждого числа на простоту не требуется.
- Временная сложность составляет `O(n log log n)`.
- Пространственная сложность составляет `O(n)`.
- При больших диапазонах работает значительно быстрее последовательной проверки чисел на простоту.

### Вывод

Решето Эратосфена позволяет найти сразу все простые числа заданного диапазона, после чего их сумма вычисляется за один проход. Благодаря этому данный подход значительно превосходит последовательную проверку каждого числа и является наиболее эффективным решением данной задачи.

In [3]:
def get_result_2(limit: int) -> int:
    is_prime = [True] * limit
    is_prime[0] = is_prime[1] = False

    for i in range(2, int(limit ** 0.5) + 1):
        if is_prime[i]:
            for j in range(i * i, limit, i):
                is_prime[j] = False

    return sum(i for i, prime in enumerate(is_prime) if prime)


start = perf_counter()
print(get_result_2(2_000_000))
end = perf_counter()

print(f'{end - start:.5f}')

142913828922
0.19564


## Сравнение времени выполнения

Для сравнения производительности используется функция `timeit()`, предназначенная специально для измерения времени выполнения небольших участков кода.

Каждый алгоритм запускается **100 раз**, после чего общее время делится на количество запусков. Таким образом получается **среднее время выполнения одного запуска**.

In [4]:
from timeit import timeit

funcs = (
    get_result_1,
    get_result_2,
)

for i, func in enumerate(funcs, start=1):
    tm = timeit(lambda: func(2_000_000), number=100) / 100
    print(f'Алгоритм {i}: {tm:.5f}')

Алгоритм 1: 1.02726
Алгоритм 2: 0.15491


## Заключение

В ходе решения задачи были рассмотрены два различных подхода к вычислению суммы всех простых чисел, меньших заданного значения.

Первый алгоритм основан на последовательной проверке каждого нечётного числа на простоту с использованием функции `isprime()` из библиотеки `sympy`. Такой подход прост в реализации и хорошо подходит для небольших диапазонов, однако требует отдельной проверки каждого числа.

Во втором решении используется **решето Эратосфена**. Вместо проверки каждого числа алгоритм заранее исключает все составные числа, после чего остаётся только просуммировать найденные простые. Благодаря этому количество выполняемых операций существенно сокращается.

Среднее время выполнения алгоритмов:

- **Алгоритм 1:** `1.02726` с
- **Алгоритм 2:** `0.15491` с

Полученные результаты показывают, что при поиске большого количества простых чисел решето Эратосфена значительно превосходит последовательную проверку каждого кандидата. В данной задаче второе решение оказалось примерно в **6 раз быстрее** первого, что делает его наиболее предпочтительным для работы с большими диапазонами.